content/1_docs/2_cookbook/3_templating/0_understanding-oop/recipe.txt

```diff
@@ -0,0 +1,462 @@
+
+If a class extends another class, we speak of inheritance, one of the advantages of OOP. A child class inherits all methods and properties of the parent class, like a real-life child inherits certain "properties" from its parents.
+
+However, in the same way that a child is not an exact copy of one of its parents, a child class can modify the inherited properties and methods and add additional ones. Inheritance is therefore a way to diversify while at the same time keeping code DRY ("Don't Repeat Yourself"). You only have to redefine or extend where necessary.
+
+But enough theory, let's just try it out with a little help of more coffee, and actually extend the `Book` class. In a new file `ebook.php` in the same folder as above, create the `Ebook` child class and add a new property and its setter and getter.
+
+```php "/book/ebook.php"
+<?php
+class Ebook extends Book
+{
+  protected $format;
+  
+  public function setFormat($format) {
+    $this->format = $format;
+  }
+
+  public function getFormat() {
+    return $this->format;
+  }
+}
+```
+
+Then in your `index.php`, require this new file and instantiate a new `Ebook` object:
+
+```php "/book/index.php"
+<?php
+ini_set('display_errors', 1);
+require_once __DIR__ . '/book.php';
+require_once __DIR__ . '/ebook.php';
+
+// … rest of code we added before
+
+$ebook = new Ebook('Form Design Patterns', 'Adam Silver');
+$ebook->setFormat('mobi');
+var_dump($ebook);
+```
+
+We can now start adding additional methods or modify the existing ones, remove the `type` property from the `Book` class because that's not needed anymore… But I leave that to you to play around with: practice makes perfect, as the saying goes.
+
+Have you ever created a Page model in Kirby? A Page model is exactly such an extended version of the original `Page` class, where you can add your custom page methods or overwrite the existing ones. 
+## Namespaces
+
+In larger applications, we often make use of third party libraries so as not having to reinvent the wheel.
+
+When we do so, it might happen that these third party libraries use the same class names as we do in our own code, which can easily result in errors, because classes like functions cannot be redeclared multiple times.
+
+To prevent such collisions, classes usually live in their own namespaces, so that even if a class has the same name, there will be no conflicts. Namespaces are also a good way to group different classes logically.
+
+Let's add a namespace to our `Book` and `Ebook` classes, for example `Cookbook\Books`:
+
+```php "/book/book.php"
+<?php
+
+namespace Cookbook\Books;
+
+class Book {
+    //… code as before
+}
+```
+
+```php "/book/ebook.php"
+<?php
+
+namespace Cookbook\Books;
+
+class EBook {
```

Comment:
The class was named `Ebook` above with lowercase `b`. Both are fine, but IMO they should be consistent. This also has an influence on the file name.